### PR TITLE
Patchwise training and inference support

### LINKED ISSUE #22: Patchwise training and inference
Some `deepsensor` users may have dense environmental data spanning large spatial areas. For example:
* Problems with data spanning the whole globe
* Problems with high-resolution satellite data

In such cases, training and inference with a `ConvNP` over the entire region of data may be computationally prohibitive. Currently, the `TaskLoader` will sample context and target data over the entire spatial region that data is available, which could produce OOM issues. So we need to support chopping the data into smaller spatial patches.

 **Training**
Supporting patchwise `ConvNP` training should just be a matter of updating the `TaskLoader` to slice the context and target datasets spatially to subsetted squares/regions before proceeding with the `TaskLoader.__call__` sampling functionality for generating `Task` objects. I believe this should be quite simple: for `xarray` data this would be `ds.sel(x1=slice(...), x2=slice(...))`, while for `pandas` data it would be `df.loc[slice(...), slice(...)]`.

**Inference**
Inference using the high-level `DeepSensorModel.predict` interface also needs support for patching. This requires functionality to stitch all the individual patch predictions together. 

For on-grid `xarray` prediction, one solution might be to call `.predict` recursively over all the patches and then concatenate the resulting `xr.Dataset`s into single objects. This would require some kind of `patchify` bool to control this and avoid infinite recursion within the inner call. Open to other ideas!

However, model predictions could differ substantially from one side of a patch border to another (due to differing context information in each patch). We therefore may need to think about having overlapping patches and averaging model predictions somehow.

**Patch size/location question**
An open question is how the size and location of the patches should be determined. One option is to have the user pass the patch size in `TaskLoader.__call__` or `DeepSensorModel.predict`, and then the location will be generated randomly unless further kwargs are passed to override this and specify exact x1/x2 spatial bounds.

### PR DESCRIPTION
**Pull request by nilsleh**
This PR aims to close #22 by implementing an option to run patch wise training.

The current approach is to expect normalized coordinates as a patch size sequence argument for the x1 and x2 dimension. The current patch size  sampling strategy is random uniform sampling.

The way I have currently thought about supporting patch wise training is the following:
- add a class method to `TaskLoader` which samples a uniform point in the normalized coordinate frame and takes the patch size to define a "bounding box" around that sampled form
- subsequently check that for the sampled box there is context and target data available, a caveat here is the while loop which is at risk of running very long if samples are scarce
- if a bbox fullfills those criteria it is used in the context and target sampling
  - for xarrays it is slice `isel` statement 
  - for df its more involved because slice expects ordered columns of the coordinates which cannot be guaranteed
- the sampled bbox is used to sample both the context and target set
- if no `patch_size` is specified to the task loader call, there are no changes, default is None so everything should run as before

TODO:
- correct visualization context encoding
- work with unnormalized coordinates
- patch wise inference that does stitching
- unit tests for sampling strategy with numpy array, as the interp coordinates are not necessarily known beforehand to respect the random sampled bounds

### CODE REVIEW

File: deepsensor/data/loader.py [hunk 1]
```diff
@@ -797,6 +829,66 @@ def sample_offgrid_aux(
             Y_t_aux = Y_t_aux.reshape(1, *Y_t_aux.shape)
         return Y_t_aux
 
+    def sample_patch_size_extent(self) -> Sequence[float]:
+        """Sample patch size.
+
+        :return sequence of patch spatial extent as [lat_min, lat_max, lon_min, lon_max]
+        """
+        # assumption of normalized spatial coordinates between 0 and 1
```
Comment (tom-andersson) [inline deepsensor/data/loader.py:837]: We can't assume data is bounded in [0, 1]. This is not guaranteed or enforced in any part of the DeepSensor data processing pipeline. Instead, we need a new method, run during the TaskLoader `init`, which computes the global min/max coordinate values of the context/target data, and then the central point of the patch should be sampled uniformly in this range.

File: deepsensor/data/loader.py [hunk 2]
```diff
@@ -1161,58 +1269,26 @@ def sample_variable(var, sampling_strat, seed):
                     f"with the `links` attribute if using the 'gapfill' sampling strategy"
                 )
```
Comment (tom-andersson) [inline deepsensor/data/loader.py]: I noticed a bunch of missing code here. I'm confused because your branch is only 2 commits behind `main` (and these 2 commits are irrelevant). Is there any chance you rejected some changes from `main` when merging?

File: deepsensor/data/loader.py [hunk 3]
```diff
@@ -881,6 +974,9 @@ def task_generation(
             "split" sampling strategy for linked context and target set pairs.
             The remaining observations are used for the target set. Default is
             0.5.
+        patch_size: Sequence[float], optional
+            Desired patch size in lat/lon used for patchwise task generation. Usefule when considering
```
Comment (tom-andersson) [inline deepsensor/data/loader.py:978]: There are a few references to lat/lon specifically. Please instead use the DeepSensor standardised coordinate names `x1`/`x2` in comments and variables. The TaskLoader operates only on standardised/normalised data.

File: deepsensor/data/loader.py [hunk 4]
```diff
@@ -1226,7 +1302,7 @@ def sample_variable(var, sampling_strat, seed):
                 X_c_offrid_all = np.concatenate(X_c_offgrid, axis=1)
             Y_c_aux = (
                 self.sample_offgrid_aux(
-                    X_c_offrid_all, self.time_slice_variable(self.aux_at_contexts, date)
+                    X_c_offrid_all, self.time_slice_variable(self.aux_at_contexts, date), sample_patch_size
```
Comment (tom-andersson) [inline deepsensor/data/loader.py:1347]: We shouldn't need to spaitally slice offrid aux; this will happen implicitly because the context data used for sampling the `self.aux_at_contexts` xarray data will already have been spatially sliced.

File: deepsensor/data/loader.py [hunk 5]
```diff
@@ -797,6 +829,66 @@ def sample_offgrid_aux(
...
+        """
+        # assumption of normalized spatial coordinates between 0 and 1
+
+        lat_extend, lon_extend = self.patch_size
+
+        lat_side = lat_extend / 2
+        lon_side = lon_extend / 2
+
+        # sample a point that satisfies the boundary and target conditions
+        continue_looking = True
```
Comment (tom-andersson) [inline deepsensor/data/loader.py:845]: I would remove the `continue_looking` logic entirely. Firstly, it's fine if the patch contains no context data; DeepSensor models should be able to handle this. The main risk here is that the patch contains no target data, which can lead to NaNs when passed to the `ConvNP.loss_fn`. However, it is much, much, easier to check for `Task`s with no target data as a training pre-processing step. This would be a separate PR or something we expect the user to be aware of.

File: deepsensor/data/loader.py [hunk 6]
```diff
@@ -1161,58 +1269,26 @@ def sample_variable(var, sampling_strat, seed):
...
-
-                    target_var = target_var.where(
-                        target_mask
-                    )  # Only keep target locations
-
-                    context_slices[context_idx] = context_var
-                    target_slices[target_idx] = target_var
+        # sample common patch size for context and target set
+        if self.patch_size is not None:
+            sample_patch_size = self.sample_patch_size_extent()
```
Comment (tom-andersson) [inline deepsensor/data/loader.py:1316]: I would suggest we don't make `patch_size` a class attribute like this; it should only exist in the scope of `__call__` here

**Review by tom-andersson — CHANGES_REQUESTED** [review]

**Comment by tom-andersson** [comment]
Hi @nilsleh, I've submitted my review. I hope you don't mind all the feedback - do let me know if you don't have the time and would prefer me to take over.

As a side note, to close https://github.com/tom-andersson/deepsensor/issues/22 we'll need to solve the 'inference' part by adding a patch-processing feature to `DeepSensorModel.predict`.

File: deepsensor/data/loader.py (hunk 2 above)
Comment (nilsleh) [reply deepsensor/data/loader.py]: Sorry, I noticed that an commited it already, but I think you were already reviewing in between, should be fixed.

File: deepsensor/data/loader.py (hunk 1 above)
Comment (nilsleh) [reply deepsensor/data/loader.py:837]: Okay, to my understanding the TaskLoader only works on already normalized/standardized data and the coordinate bounds were normalized to [0,1] but that is good to know, thanks!

File: deepsensor/data/loader.py (hunk 5 above)
Comment (nilsleh) [reply deepsensor/data/loader.py:845]: Is there an assumption about a common coordinate range between the context and the target? Because if so, we can gather the coordinate bound extend of the target variable and use that to do the random window sampling?

Comment (tom-andersson) [reply deepsensor/data/loader.py:845]: No, unfortunately we can't assume that. We'll have to loop over all the `self.context` and `self.target` variables updating the min/max data coordinate bounds.

File: deepsensor/data/loader.py (hunk 1 above)
Comment (tom-andersson) [reply deepsensor/data/loader.py:837]: By default, the `DataProcessor` linearly normalises the coords of the first data variable it is provided with to lie in `[0, 1]`, but subsequent variables may exceed that data range. Thus, although the data coords will typically lie in `[0, 1]`, there is nothing constraining this to always hold.

File: deepsensor/data/loader.py [hunk 7]
```diff
@@ -1279,24 +1487,113 @@ def __call__(
...
-                    target_sampling,
-                    split_frac,
-                    datewise_deterministic,
-                    seed_override,
+        assert patch_strategy in [None, "random", "sliding"], (
+            f"Invalid patch strategy {patch_strategy}. "
+            f"Must be one of [None, 'random', 'sliding']."
+        )
+
+        if patch_strategy is None:
```
Comment (nilsleh) [inline deepsensor/data/loader.py:1495]: I moved the logic to the `_call_` function, however, there is quiet a bit of code redundancy  because:
- checking separate sampling strategies
- checking whether one supplies a single data or a sequence of date that determines whether a `Task` or a `list[Task]` is returned

So that can be made more concise

File: tests/test_task_loader.py [hunk 8]
```diff
@@ -270,6 +270,103 @@ def test_links(self) -> None:
...
+
+        context = [da_data_0_1, da_data_smaller, da_data_larger]
+        tl = TaskLoader(
+            context=context,  # gridded xarray and off-grid pandas contexts
+            target=self.df,  # off-grid pandas targets
+        )
+
+        # TODO it would be better to do this with pytest.fixtures
+        # but could not get to work so far
+        task = tl(
```
Comment (nilsleh) [inline tests/test_task_loader.py:306]: It would be better to have fixtures that generate the data setup and then we can test with different configurations like single date, list of dates, different context and target sampling strategies  etc.

File: tests/test_training.py [hunk 9]
```diff
@@ -114,6 +115,64 @@ def test_training(self):
...
+        epoch_losses = []
+        for epoch in tqdm(range(n_epochs)):
+            batch_losses = trainer(train_tasks, batch_size=batch_size)
+            epoch_losses.append(np.mean(batch_losses))
+
+        # Check for NaNs in the loss
+        loss = np.mean(epoch_losses)
+        self.assertFalse(np.isnan(loss))
+
+    def test_sliding_window_training(self):
```
Comment (nilsleh) [inline tests/test_training.py:152]: I think the trainer tests could also benefit from fixtures, because at the moment there is a lot of code duplication. Maybe fixtures that generate a list of `train_tasks` that are then run with a `ConvCNP` and `Trainer`

File: deepsensor/data/loader.py (hunk 2 above)
Comment (davidwilby) [reply deepsensor/data/loader.py:1297]: Is the below section in this diff the missing section? @nilsleh could you help me understand what's happening to the below functionality please? Has it been moved somewhere, is it no longer required? Thanks.

Comment (davidwilby) [reply deepsensor/data/loader.py:1297]: @nilsleh ?

**Comment by davidwilby** [comment]
Noting here that, among other things, we need to work on `Task.__repr__` and `Task.__str__` to handle patched tasks and return meaningful output (for instance when running `print(task)`).

At present, `print`ing a patched task results in one set of output for each patch, which is unhelpful.

`__str__` and `__repr__` are also caused to iterate through `bbox`, which contain floats, resulting in an error since floats don't have `.dtype` or `.shape` methods.

https://github.com/alan-turing-institute/deepsensor/blob/7ec755fa4edab2744ee94af4aa20021ef83a4a82/deepsensor/data/task.py#L59-L64

I also notice that `bbox` seems to include a mix of `float` and `float64`, which seems odd.

File: deepsensor/data/loader.py (hunk 2 above)
Comment (davidwilby) [reply deepsensor/data/loader.py]: Resolved by [f5e4a8a](https://github.com/alan-turing-institute/deepsensor/pull/75/commits/f5e4a8aa4cf36ff7cdd8291c558d8ca70c448178)
